Background jobs

aapeliv opened the pull request:
This is the start of a background job system.

Main TODO:

* ~Overall implementation needs a bit of mulling over~
* ~Scheduled tasks (e.g. send notifications about messages every 5 min, purge leftover tokens once a day, etc)~
* ~Tests~
* ~The fg and bg stuff is kind of messy, and they're all fighting for the same GIL, this can probably be improved~

aapeliv committed:
Implement a very crude background job system

aapeliv committed:
Write process_job and clean stuff up

amdwit commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,61 @@
+            return
+
+        # mark it as ours
+        job.state = BackgroundJobState.working
+        job.try_count += 1
+        session.commit()
```
Is it possible that at the same moment another `process_job` is executing these same lines for the same job because two `service_jobs` functions selected the same job?

aapeliv replied:
Yes, it's broken at the moment

lucaslcode commented on app/backend/src/couchers/models.py:
```diff
@@ -654,3 +654,60 @@ class InitiatedUpload(Base):
         return (self.created <= func.now()) & (self.expiry >= func.now())
+
+
+class BackgroundJobType(enum.Enum):
```
Is it time to create a models folder? :P

amdwit replied:
Great idea! I just created a draft for it: https://github.com/Couchers-org/couchers/pull/482

kalvdans commented:
No mail queues in the database please. Use a separate MTA for the mail queues, as discussed in https://github.com/Couchers-org/couchers/issues/282.

aapeliv commented:
> No mail queues in the database please. Use a separate MTA for the mail queues, as discussed in #282.

Let's discuss on #282, see my comment there

kalvdans commented on app/backend/src/couchers/jobs/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from couchers.jobs.email import process_send_email
```
No code in `__init__.py` please.

aapeliv replied:
OK, I'll fix it

kalvdans commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,61 @@
+        session.commit()
+
+
+def service_jobs():
```
I think it is an easier design to have a cron job, that checks for stuff to do and then exits. Instead of having a daemon process.

kalvdans commented on app/pb/jobs.proto:
```diff
@@ -0,0 +1,13 @@
+// background job stuff, not an API
```
so don't make a `.proto`?

kalvdans reviewed (changes requested):
I would like the email reminders to be handled by a cron job, and the database shuold remember things like "last time a reminder was sent to user X" instead of an explicit job queue.

kalvdans commented on app/backend/src/app_fg.py:
```diff
@@ -0,0 +1,95 @@
+
+def wait_for_foreground(fg):
+    server, jailed_server, open_server, media_server = fg
+    server.wait_for_termination()
+    jailed_server.wait_for_termination()
+    open_server.wait_for_termination()
+    media_server.wait_for_termination()
```
These services should never die gracefully. [Crash-only software](https://en.wikipedia.org/wiki/Crash-only_software) for the win!

aapeliv commented:
> I would like the email reminders to be handled by a cron job, and the database shuold remember things like "last time a reminder was sent to user X" instead of an explicit job queue.

Yes, I agree that email reminders ought to be done by a scheduled job that just checks the database for which reminders need sending.

aapeliv commented:
I think the main question here is whether to do scheduling within python or pull in another dependency (e.g. `cron`). In my experience, cron does not play very nicely with Docker, which is unfortunate. I'm also, again, not keen on pulling in another dependency.

One more thing is that we will need background jobs (non-scheduled, run basically straight after invoked, and run in the background) at some point.

aapeliv committed:
Fix up some review comments

aapeliv committed:
Fix concurrency issues

aapeliv committed:
Move internal protobufs into different folder

aapeliv committed:
Clean up

aapeliv committed:
Fix tests

aapeliv committed:
Implement scheduled jobs

aapeliv committed:
Submit jobs to futures.ThreadPoolExecutor

aapeliv committed:
Work on concurrency

aapeliv committed:
Work on fixing threading issues

aapeliv committed:
Fix issues with forks sharing connections

aapeliv committed:
Fix up job scheudler and clean up

aapeliv committed:
Restart workers on exceptions

aapeliv committed:
Fix up scheduled jobs

aapeliv committed:
Fix models

aapeliv committed:
Fix migrations

aapeliv committed:
Add a few light tests for bg jobs

aapeliv committed:
Add src/couchers/jobs/__init__.py for coverage

aapeliv committed:
Add test for login token pruge job

aapeliv committed:
Add test for service_jobs

aapeliv committed:
Get out of infinite loop in a nicer way

aapeliv committed:
Clean up

aapeliv committed:
Add signup token purging and tests

aapeliv committed:
Add test for scheduler

aapeliv committed:
Fix tests

aapeliv committed:
Fix up ready_for_retry and add job retry test

aapeliv committed:
Clean up

aapeliv committed:
Add scheduler continuation test

aapeliv committed:
Add some more assertions

aapeliv committed:
Regenerate migrations

aapeliv commented:
@kalvdans: I've added tests to this, could you review, please?

kalvdans commented on app/backend/src/couchers/db.py:
```diff
@@ -3,11 +3,13 @@
 import logging
 import os
 import re
+import threading
```
Unused include

kalvdans commented on app/backend/src/couchers/email/dev.py:
```diff
@@ -1,13 +1,16 @@
+logger = logging.getLogger(__name__)
+
 
 def print_dev_email(sender_name, sender_email, recipient, subject, plain, html):
```
Please describe what this function returns in a docstring.

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1044,3 +1044,82 @@ class Reply(Base):
+    max_tries = Column(Integer, nullable=False, default=5)
+
+    # protobuf encoded job payload
+    payload = Column(Binary, nullable=False)
```
How do we handle database upgrades when the protobuf encoding is changed? Nuke the queue?

aapeliv replied:
Yes, either nuke the queue, or only make backwards compatible changes to the protobuf. The queue should never contain anything particularly long-lived.

kalvdans commented on app/backend/src/tests/test_bg_jobs.py:
```diff
@@ -0,0 +1,332 @@
+        _run_job_and_schedule(sched, schedule_id)
+
+    with pytest.raises(EndOfTime):
+        with patch("couchers.jobs.worker._run_job_and_schedule", mock_run_job_and_schedule):
+            with patch("couchers.jobs.worker.SCHEDULE", MOCK_SCHEDULE):
+                with patch("couchers.jobs.worker.monotonic", mock_monotonic):
+                    with patch("couchers.jobs.worker.sleep", mock_sleep):
```
I prefer to use the `monkeypatch` pytest fixture. You use it like `monkeypatch.setattr(couchers.jobs.worker, "_run_job_and_schedule", mock_run_job_and_schedule)` and it has effect until the end of the test so no indentation needed.

aapeliv replied:
Thanks, that's much cleaner!

kalvdans commented on app/pb/internal/jobs.proto:
```diff
@@ -0,0 +1,13 @@
+// background job stuff, not an API
```
move to `app/backend` and don't generate javascript wrappers for it.

aapeliv replied:
Yeah, that's cleaner. Thanks.

kalvdans commented on app/backend/src/couchers/email/__init__.py:
```diff
@@ -72,16 +73,24 @@ def _render_email(template_file, template_args={}):
+    )
 
 
 def send_email_template(recipient, template_file, template_args={}):
```
Rename to `enqueue_email_with_template` since it doesn't send any email.

kalvdans commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,177 @@
+
+
+def process_job():
+    logger.info(f"Processing a job")
```
"Looking for a job to grab"

kalvdans commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,177 @@
+    with session_scope(isolation_level="REPEATABLE READ") as session:
+        # a combination of REPEATABLE READ and SELECT ... FOR UPDATE SKIP LOCKED makes sure that only one transaction
+        # will modify the job at a time. SKIP UPDATE means that if the job is locked, then we ignore that row, it's
+        # easier to use SKIP LOCKED vs NOWAIT in the ORM, with NOWAIT you get an ugly error from psycopg2
```
Insert reference to psycopg2 issue report.

aapeliv replied:
It's not a bug, just an annoying exception, and sqlalchemy makes it a bit annoying to catch just certain psycopg2 exceptions. I've updated the comment to clarify.

kalvdans commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,177 @@
+            ret = func(message_type.FromString(job.payload))
+            job.state = BackgroundJobState.completed
+            logger.info(f"Job #{job.id} complete on try number {job.try_count}")
+        except Exception as e:
```
Please only catch errors happening during the `func` call. KeyError from the lookup in JOBS is fatal and should kill the worker.

aapeliv replied:
Hmmm, alright. I was first thinking that the keyerror ought to be caught as well, but can move it outside too.

kalvdans commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,177 @@
+            ret = func(message_type.FromString(job.payload))
+            job.state = BackgroundJobState.completed
+            logger.info(f"Job #{job.id} complete on try number {job.try_count}")
+        except Exception as e:
```
Would it be possible to narrow down the specific exceptions we expect to happen? OSError?

aapeliv replied:
I'm not sure. Maybe we can exclude some exceptions?

kalvdans replied:
Let's postpone for later.

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1044,3 +1044,82 @@ class Reply(Base):
+    # job is fresh, waiting to be picked off the queue
+    pending = 1
+    # job has been grabbed by a worker, and is being worked on
+    reserved = 2
```
Unused status, as the row is locked instead.

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1044,3 +1044,82 @@ class Reply(Base):
+
+class BackgroundJobType(enum.Enum):
+    # payload: org.couchers.internal.jobs.SendEmailPayload
+    send_email = 1
```
Could use `= enum.auto()` to show that the actual value doesn't matter.

aapeliv replied:
This is basically the case with every enum in the db, right?

kalvdans replied:
Will be done in a separate PR, resolving!

kalvdans commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,177 @@
+    """
+    Schedules jobs according to schedule in .definitions
+    """
+    # multiprocessing uses fork() which in turn copies file descriptors, so the engine may have connections in its pool
```
So don't connect to the engine before forking.

aapeliv replied:
At the moment the app checks the db for a connection before basically doing anything, and fails fast if it can't ping the DB. (This happens in app.py). It's much easier to catch that way.

I think it's less error prone to keep these in here, and it took me quite a while to figure it out (hence the big comment...).

kalvdans replied:
We can remove this if we later refactor to fork in a shell script. Resolving.

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1044,3 +1044,82 @@ class Reply(Base):
+class BackgroundJobType(enum.Enum):
+    # payload: org.couchers.internal.jobs.SendEmailPayload
+    send_email = 1
+    # payload: google.protobuf.Empty
+    purge_login_tokens = 2
+    # payload: google.protobuf.Empty
+    purge_signup_tokens = 3
```
These two are periodical events, living in RepeatedJob, and thus have no payload, right?

aapeliv replied:
That's right. But `google.protobuf.Empty` is just an empty message.

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1044,3 +1044,82 @@ class Reply(Base):
+class BackgroundJob(Base):
+    """
+    This table implements a queue of background jobs.
+    """
```
I don't see rows deleted from this table ever. Is the plan to have a background job to purge old background jobs :inception: :smile:

aapeliv replied:
Yeah, I guess... Could add a job for that!

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1044,3 +1044,82 @@ class Reply(Base):
+
+    id = Column(BigInteger, primary_key=True)
+
+    job_type = Column(Enum(BackgroundJobType), nullable=False, unique=True)
```
Would be nice to split the BackgroundJobType into repeatable jobs and one-time jobs.

aapeliv replied:
I think it's cleaner to have them together. The repeatable jobs just have a property that they're scheduled regularly by another process. That should not change how they're processed.

It also allows us to easily run otherwise repeatable jobs on-demand (e.g. if there is some kind of report, it can be done on schedule or triggered manually), and basically use them interchangably.

What do you see as the advantage of splitting?

kalvdans replied:
> It also allows us to easily run otherwise repeatable jobs on-demand 

Isn't it easier to just run the Python code without querying the database in the first place then?

aapeliv replied:
> Isn't it easier to just run the Python code without querying the database in the first place then?

Do you mean running the code in the foreground? The non-repeated jobs are basically there for async background processing, so the user does not need to wait several seconds for a bunch of background cogs to turn.

kalvdans replied:
The repetetive jobs `process_purge_login_tokens` and `process_purge_signup_tokens` are idempotent and can run as often or seldom as needed without harm. Therefore they don't need to store the last time they were run in the database.

kalvdans commented on app/backend/src/app_bg.py:
```diff
@@ -0,0 +1,12 @@
+import logging
```
Remove this file, call start_job_servicer directly instead.

aapeliv replied:
Resolved?

kalvdans replied:
Yep, looks good.

kalvdans commented on app/backend/src/couchers/db.py:
```diff
@@ -35,16 +37,30 @@ def apply_migrations():
+    """
+    Creates an engine with the given isolation level.
+
+    Do not change isolation_level unless you know what you're doing!
```
Useless scary comment, remove or replace with "Normally you should use the default READ COMMITTED which will make all changes in a transaction visible at the same time, but no other atomicity guarantees"

kalvdans commented on app/backend/src/couchers/db.py:
```diff
@@ -35,16 +37,30 @@ def apply_migrations():
         return create_engine(config.config["DATABASE_CONNECTION_STRING"])
 
 
+@functools.cache
```
don't need to cache if it returns a shallow copy.

kalvdans commented on app/backend/src/couchers/config.py:
```diff
@@ -11,6 +11,8 @@
 CONFIG_OPTIONS = [
     # Whether we're in dev mode
     ("DEV", bool),
+    # Whether we're fg (answering API queries) or bg processing
+    ("ROLE", ["fg", "bg", "both"], "both"),
```
I would prefer a shell script starting the two services and backgrounding one of them with `&`, instead of all the boilerplate in Python to create two processes.

aapeliv replied:
Let's discuss in the meeting today.

It's kind of messy at the moment, I'll try to clean it up a tad bit.

The bottom line is that I'd like it to be as simple as possible for new devs to spin it up, and that means having only one backend container running. But we can have config to run them separately and should do that in prod.

kalvdans replied:
It is perfectly fine to have a shell script as an entry point in a container. See you tonight!

kalvdans replied:
The outcome of the meeting is that it doesn't really matter if we fork in Python or in shell so we could leave it as it stands. Resolving conversation.

kalvdans reviewed (changes requested):
Great with tests! :thumbsup:

aapeliv committed:
Fix a whole lot of review comments

aapeliv committed:
Move backend-specific proto stuff out of global pb/ folder

aapeliv committed:
Clean up

kalvdans reviewed (changes requested):
I closed off some of my earlier comments but some remains unanswered.

aapeliv commented:
@kalvdans: would you like me to get rid of the `app_{fg,bg}.py` and basically just move those bits of code back into `app.py`? If you wish, I can also inline `start_job_servicer` too?

kalvdans commented:
> @kalvdans: would you like me to get rid of the `app_{fg,bg}.py` and basically just move those bits of code back into `app.py`?

Basically I would like the choice of apiservicer/emailworker/both to be controlled by different start scripts *or* command-line arguments. But I am fighting uphill since docker-compose is so obsessed with environment variables, so I give up :) Go with your code.

aapeliv committed:
Clean up

aapeliv committed:
Fix migrations

aapeliv committed:
Fix up app.py

kalvdans commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,183 @@
+
+
+def service_jobs():
+    """
+    Don't run many of these simultaneously: it'll work but it'll conflict and cause a lot of missed job processing.
+    """
```
Why? The row locking should be solid.

aapeliv replied:
Yeah, row locking itself is solid in that jobs themselves will be run only once. But the scheduler checks how many jobs there are and launches a worker for each. If there's multiple schedulers they might launch twice the workers causing half the workers to not acquire a job and exit. Just unnecessarily burdening the DB.

Of course this depends only really on the amount of time it takes to check how many pending jobs there are vs how that worker starting and attempting to acquire a job. If that time is very small (which it should really be), and there are multiple schedulers, they shouldn't really interfere with each other.

kalvdans replied:
If the worker threads are slow to grab a job from the database, then the main thread will keep submitting new callbacks to the ThreadPoolExecutor, possibly making a bad situation worse.

This is unnecessary complex IMHO. Just start all the workers immediately and let them poll the database themselves.

aapeliv replied:
Hmmm, that's a good point.

I was hoping to eventually move to a pg `LISTEN ... NOTIFY` system to spawn workers based on actual events, and run the scheduler only to clean up any jobs that were left over in case. That's what inspired this style.

I agree it's more complex now, but I think now it's pretty solid. It doesn't really matter now though. I can simplify it though if you think it's better. (I would prefer to keep `process_job` mostly intact though, and just have the `service_jobs` function call it directly, waiting a second if it got no job)

kalvdans replied:
> I would prefer to keep process_job mostly intact though, and just have the service_jobs function call it directly, waiting a second if it got no job

This sounds good. `process_job` is good to keep as a function, releasing the lock automatically when it returns or throw exception.

aapeliv replied:
Simplified, what do you think?

kalvdans replied:
Much better! We might need to adjust the log levels of the spamming log messages but that's details. Resolving conversation!

aapeliv committed:
Simplify service_jobs

aapeliv commented:
(I'm not sure why I can't reply in thread.)

> The repetetive jobs process_purge_login_tokens and process_purge_signup_tokens are idempotent and can run as often or seldom as needed without harm. Therefore they don't need to store the last time they were run in the database.

I don't think we want to make enforce this idempotency requirement. In particular, I think there are some jobs that we'd like to run no more than at a certain frequency, though this doesn't guarantee that it won't take longer. I think we want to soon be doing jobs that run at regular times, this is step towards that, I suppose. But I guess you could say YAGNI...

kalvdans commented:
> (I'm not sure why I can't reply in thread.)

Because I bundled two indipendent replies together with a review. I never understand the logic behind github, gitlab treats them as if they were replied to one by one.

> > The repetetive jobs process_purge_login_tokens and process_purge_signup_tokens are idempotent and can run as often or seldom as needed without harm. Therefore they don't need to store the last time they were run in the database.
> 
> I don't think we want to make enforce this idempotency requirement. In particular, I think there are some jobs that we'd like to run no more than at a certain frequency, though this doesn't guarantee that it won't take longer. I think we want to soon be doing jobs that run at regular times, this is step towards that, I suppose. But I guess you could say YAGNI...

Yeah, I get the feeling that you have a bigger picture in your head. But it is hard to review then... If you are confident it is the right thing in the long run, it is fine with me. I'll review once all job types are in.

aapeliv commented:
> Yeah, I get the feeling that you have a bigger picture in your head. But it is hard to review then... If you are confident it is the right thing in the long run, it is fine with me. I'll review once all job types are in.

Maybe I'm thinking wrong.

Eventually we will want some jobs to run at the same time every day, or at multiple points. Yes... like cron...

I have three other job types in mind at the moment:

1. A simple notification job for emails about unseen messages. Runs every 5 minutes and sends emails for any unseen messages more than 15 minutes old, or similar. It's not entirely clear to me how this will look. But there should be at most one job running at a time (to not send duplicate emails).
2. Sending notifications about events coming up or other "daily emails" in each user according to morning in their timezone (or similar).
3. A job to create reports for the demanding Itsi. I'm not sure what this entails, some possibly longer-running SQL queries. It would be nice I suppose for these to be run the same time every day, but some drift is not a big problem.

But maybe all of these should really be idempotent anyway (i.e. track their own last-run status in whatever form).

I'm not sure, what do you think?

I'm happy to shrink this PR to best suit the currently implemented features, if you think anything is adding unnecessary bloat, etc.

kalvdans commented:
> But maybe all of these should really be idempotent anyway (i.e. track their own last-run status in whatever form).

I would very much prefer that style!

aapeliv committed:
Remove RepeatedJob state from database

aapeliv committed:
Remove restart test as it no longer works

It was broken anyway, the code should have been deindented after the first pytest.raises with block.

aapeliv commented:
> I would very much prefer that style!

I've removed the repeated job state from the database. Repeated jobs now have to be idempotent and the scheduler will schedule each one initially at startup, then following the schedule.

Please re-review and point out anything else required for a merge! :)

aapeliv committed:
Fix migrations after rebase

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1068,3 +1068,65 @@ class Reply(Base):
+class BackgroundJobType(enum.Enum):
+    # payload: jobs.SendEmailPayload
+    send_email = 1
+    # payload: google.protobuf.Empty
+    purge_login_tokens = 2
+    # payload: google.protobuf.Empty
+    purge_signup_tokens = 3
```
Remove repeated jobs from database enum.

aapeliv replied:
Could they not still be run through the background queue?

kalvdans replied:
You can still run the jobs every 24 hours, just don't add them into in the database schema please, since they already store the state they need in the other tables.

aapeliv replied:
The abstraction here is that a background job is some longer-running or async operation that should be retried. It takes a certain protobuf as an input and produces no output. A failing job is retried a few times with exponential backoff. Background jobs are serviced by a set of workers asynchronously from a job queue. The tasks themselves are defined in the same file and the job type to function lookup is centralized.

Some jobs are run as a response to user input, e.g. sending notifications or links.

Some jobs are additionally run according to some rough schedule, but may also be triggered manually.

I'd like this scheme. I'd like the repeated jobs to also take advantage of the other features of background jobs.

Also I'd really not feel like rewriting all the tests as I used the two background jobs as generic jobs (since they're idempotent and shouldn't fail) and so most tests utilize those (I need at least two jobs for testing).

I understand this model has its downsides, and depending on how the world unfolds, we probably need to add functionality and modify it as we go. But my opinion is that it's pretty good for now.

I don't see what we get from separating the repeating jobs from the non-repeating jobs, except a small decrease in complexity. But then we have two job systems: repeating and non-repeating. It won't make sense (and will be confusing) to have all the tasks in the same file if they don't behave the same, so they should be split up.

kalvdans replied:
I disagree. I my world each job keep tracks of its own state independent of other jobs. The interval each job is run at is stored in the Python source code and not in the database. Also in my world, token purging never fails :)

But as you say we could always change it later. I'll make a final codewise review of the PR!

aapeliv replied:
I will merge this, but let's have a brief chat again in the backend meeting :)

kalvdans commented on app/backend/src/couchers/jobs/worker.py:
```diff
@@ -0,0 +1,147 @@
+            logger.critical("Background worker starting")
+            func()
+        except Exception as e:
+            logger.critical("Unhandled exception in background worker", exc_info=e)
```
Perhaps a grace period here to not consume 100% cpu and keep hammering the database in case of a programming error. Remove the loop or put a `time.sleep(60)` here.

kalvdans reviewed (dismissed):
Code looks ok! Just a minor comment, feel free to merge with or without fixing it. Hereby auto-approving future commits!

aapeliv committed:
Add delay on failed worker
